core/src/main/java/org/elasticsearch/search/builder/NewSearchSourceBuilder.java

```diff
@@ -0,0 +1,1077 @@
+        boolean hasFieldNames = fieldNames != null;
+        out.writeBoolean(hasFieldNames);
+        if (hasFieldNames) {
+            out.writeVInt(fieldNames.size());
+            for (String field : fieldNames) {
+                out.writeString(field);
+            }
+        }
+        out.writeVInt(from);
+        boolean hasHighlightBuilder = highlightBuilder != null;
+        out.writeBoolean(hasHighlightBuilder);
+        if (hasHighlightBuilder) {
+            out.writeBytesReference(highlightBuilder);
+        }
+        boolean hasIndexBoost = indexBoost != null;
+        out.writeBoolean(hasIndexBoost);
+        if (hasIndexBoost) {
+            out.writeVInt(indexBoost.size());
+            for (ObjectCursor<String> key : indexBoost.keys()) {
+                out.writeString(key.value);
+                out.writeFloat(indexBoost.get(key.value));
+            }
+        }
+        boolean hasInnerHitsBuilder = innerHitsBuilder != null;
+        out.writeBoolean(hasInnerHitsBuilder);
+        if (hasInnerHitsBuilder) {
+            out.writeBytesReference(innerHitsBuilder);
+        }
+        boolean hasMinScore = minScore != null;
+        out.writeBoolean(hasMinScore);
+        if (hasMinScore) {
+            out.writeFloat(minScore);
+        }
+        boolean hasPostQuery = postQueryBuilder != null;
+        out.writeBoolean(hasPostQuery);
+        if (hasPostQuery) {
+            postQueryBuilder.writeTo(out);
+        }
+        boolean hasQuery = queryBuilder != null;
+        out.writeBoolean(hasQuery);
+        if (hasQuery) {
+            queryBuilder.writeTo(out);
+        }
+        boolean hasRescoreBuilders = rescoreBuilders != null;
+        out.writeBoolean(hasRescoreBuilders);
+        if (hasRescoreBuilders) {
+            out.writeVInt(rescoreBuilders.size());
+            for (BytesReference rescoreBuilder : rescoreBuilders) {
+                out.writeBytesReference(rescoreBuilder);
+            }
+        }
+        boolean hasScriptFields = scriptFields != null;
+        out.writeBoolean(hasScriptFields);
+        if (hasScriptFields) {
+            out.writeVInt(scriptFields.size());
+            for (ScriptField scriptField : scriptFields) {
+                scriptField.writeTo(out);
+            }
+        }
+        out.writeVInt(size);
+        boolean hasSorts = sorts != null;
+        out.writeBoolean(hasSorts);
+        if (hasSorts) {
+            out.writeVInt(sorts.size());
+            for (BytesReference sort : sorts) {
+                out.writeBytesReference(sort);
+            }
+        }
+        boolean hasStats = stats != null;
+        out.writeBoolean(hasStats);
+        if (hasStats) {
+            out.writeStringArray(stats);
+        }
+        boolean hasSuggestBuilder = suggestBuilder != null;
+        out.writeBoolean(hasSuggestBuilder);
+        if (hasSuggestBuilder) {
+            out.writeBytesReference(suggestBuilder);
+        }
+        out.writeVInt(terminateAfter);
```

Comment:
Looking at `timoueInMillis`, its not actually always positive as the default is `-1` so I'll change that to serialise as a normal long value.

For `terminateAfter` I'll add a check to the setter as I think we should stop the illegal value at the point that it is set rather than when it is written to the stream.